src/control/client/pool.go

```diff
@@ -64,11 +64,15 @@ func (c *connList) PoolCreate(req *PoolCreateReq) (*PoolCreateResp, error) {
 		return nil, err
 	}
 
-	poolUUID, err := uuid.NewRandom()
-	if err != nil {
-		return nil, errors.Wrap(err, "generating pool uuid")
+	poolUUIDStr := req.UUID
+
+	if req.UUID == "" {
+		poolUUID, err := uuid.NewRandom()
+		if err != nil {
+			return nil, errors.Wrap(err, "generating pool uuid")
+		}
+		poolUUIDStr = poolUUID.String()
 	}
```

Comment:
I would prefer to validate the supplied UUID out here in the control plane rather than letting a potentially wacky string get all the way into the ioserver.

I suggest doing it something like this instead:
```go
        poolUUID, err := uuid.Parse(req.UUID)
        if err != nil {
                if req.UUID == "" {
                        poolUUID, err = uuid.NewRandom()
                }
                if err != nil {
                        return nil, errors.Wrapf(err, "bad pool UUID: %q", req.UUID)
                }
        }
```
It's true that in the event you get an error from `uuid.NewRandom()` the error message won't be completely accurate, but looking at the code it seems like one of those "can't happen" errors and it only returns an error because it uses `uuid.NewRandomFromReader()` with an internal rand source under the covers.

Reply:
Wouldn't it be better to put the validation in an else so we can have clearer error messages?

Reply:
Up to you. I don't think you're ever going to get the other error unless the machine is completely screwed up. Mostly I'm just focused on making sure we validate the input as close to the edge as possible.